File: backend/src/domain/repository/IAuthRepository.ts
import { User } from '../entities/User';
import { GlobalProfile } from '../entities/GlobalProfile';
import { GlobalIdentityVerification } from '../entities/GlobalIdentityVerification';

/**
 * Repository interface for authentication and authorization operations
 */
export interface IAuthRepository {
  // User Registration & Login
  createUser(email: string, passwordHash: string): Promise<User>;
  findUserByEmail(email: string): Promise<User | null>;
  findUserById(userId: number): Promise<User | null>;
  updatePassword(userId: number, newPasswordHash: string): Promise<User>;

  // Profile Management
  createGlobalProfile(
    userId: number,
    profileData: Partial<GlobalProfile>,
  ): Promise<GlobalProfile>;
  findGlobalProfileByUserId(userId: number): Promise<GlobalProfile | null>;
  updateGlobalProfile(
    userId: number,
    profileData: Partial<GlobalProfile>,
  ): Promise<GlobalProfile>;

  // Identity Verification
  createIdentityVerification(
    userId: number,
    verificationData: Partial<GlobalIdentityVerification>,
  ): Promise<GlobalIdentityVerification>;
  findIdentityVerificationByUserId(
    userId: number,
  ): Promise<GlobalIdentityVerification | null>;
  findIdentityVerificationByCitizenId(
    citizenId: string,
  ): Promise<GlobalIdentityVerification | null>;

  // Account Status
  activateUser(userId: number): Promise<User>;
  deactivateUser(userId: number): Promise<User>;

  // Password Reset
  updatePasswordResetToken(
    userId: number,
    token: string,
    expiresAt: Date,
  ): Promise<boolean>;
  findUserByResetToken(token: string): Promise<User | null>;
  clearPasswordResetToken(userId: number): Promise<boolean>;

  // Email Verification
  updateEmailVerificationStatus(
    userId: number,
    isVerified: boolean,
  ): Promise<User>;
}

File: backend_java/src/main/java/com/service/backend/domain/repository/IAuthRepository.java
package com.service.backend.domain.repository;

import com.service.backend.domain.entity.User;
import com.service.backend.domain.entity.GlobalProfile;
import com.service.backend.domain.entity.GlobalIdentityVerification;
import reactor.core.publisher.Mono;

import java.time.LocalDateTime;

/**
 * Repository interface for authentication and authorization operations
 */
public interface IAuthRepository {

    // User Registration & Login
    Mono<User> createUser(String email, String passwordHash);
    Mono<User> findUserByEmail(String email);
    Mono<User> findUserById(Long userId);
    Mono<User> updatePassword(Long userId, String newPasswordHash);

    // Profile Management
    Mono<GlobalProfile> createGlobalProfile(Long userId, GlobalProfile profileData);
    Mono<GlobalProfile> findGlobalProfileByUserId(Long userId);
    Mono<GlobalProfile> updateGlobalProfile(Long userId, GlobalProfile profileData);

    // Identity Verification
    Mono<GlobalIdentityVerification> createIdentityVerification(Long userId, GlobalIdentityVerification verificationData);
    Mono<GlobalIdentityVerification> findIdentityVerificationByUserId(Long userId);
    Mono<GlobalIdentityVerification> findIdentityVerificationByCitizenId(String citizenId);

    // Account Status
    Mono<User> activateUser(Long userId);
    Mono<User> deactivateUser(Long userId);

    // Password Reset
    Mono<Boolean> updatePasswordResetToken(Long userId, String token, LocalDateTime expiresAt);
    Mono<User> findUserByResetToken(String token);
    Mono<Boolean> clearPasswordResetToken(Long userId);

    // Email Verification
    Mono<User> updateEmailVerificationStatus(Long userId, Boolean isVerified);
}
